pytd/query_engine.py

```diff
@@ -126,43 +126,88 @@ class PrestoQueryEngine(QueryEngine):
 
     def __init__(self, apikey, endpoint, database, header):
         super(PrestoQueryEngine, self).__init__(apikey, endpoint, database, header)
-        self.engine = self._connect()
+        self.prestodb_connection, self.tdclient_connection = self._connect()
 
     @property
     def user_agent(self):
         """User agent passed to a Presto connection.
         """
-        return "pytd/{0} (prestodb/{1})".format(__version__, prestodb.__version__)
+        return "pytd/{0} (prestodb/{1}; tdclient/{2})".format(
+            __version__, prestodb.__version__, tdclient.__version__
+        )
 
-    def cursor(self):
+    @property
+    def presto_api_host(self):
+        """Presto API host obtained from ``TD_PRESTO_API`` env variable or
+        inferred from Treasure Data REST API endpoint.
+        """
+        return os.getenv(
+            "TD_PRESTO_API", urlparse(self.endpoint).netloc.replace("api", "api-presto")
+        )
+
+    def cursor(self, **kwargs):
         """Get cursor defined by DB-API.
 
         Returns
         -------
-        prestodb.dbapi.Cursor
+        prestodb.dbapi.Cursor, or tdclient.cursor.Cursor
         """
-        return self.engine.cursor()
+        if len(kwargs) == 0:
+            return self.prestodb_connection.cursor()
+
+        logger.warning(
+            "returning `tdclient.cursor.Cursor`. This cursor, `Cursor#fetchone` "
+            "in particular, might behave different from your expectation, "
+            "because it actually executes a job on Treasure Data and fetches all "
+            "records at once from the job result."
+        )
+
+        original_cursor_kwargs = self.tdclient_connection._cursor_kwargs.copy()
```

Comment:
This workaround is hard to understand just looking the code. Nice to have a comment the reason for noting td-client-python behavior.

Reply:
Or, might be better to create a method since `HiveQueryEngine` also uses this workaround.

Reply:
0734a4764244b3fe74107a8cfe86987c727e0652